src/v/cluster/log_eviction_stm.cc

```diff
@@ -143,6 +143,14 @@ ss::future<> log_eviction_stm::monitor_log_eviction() {
     while (!_as.abort_requested()) {
         try {
             auto eviction_offset = co_await storage_eviction_event();
+            if (!_raft->log()->config().is_collectable()) {
```

Comment:
I stared at this for a while wondering if we need to block the new behavior on a feature flag (see `feature_table::feature`), given it's conceivable some replicas of compacted partitions in older versions wouldn't have the stm, but newer versions would during a rolling upgrade.

I haven't been able to think of a bad scenario that would really cause problems, and frankly I'm not sure there is even a place we could block with such a feature flag (see `feature_table::await_feature`).

No action required on your end, just a passing thought.